### PR TITLE
[softcap] Apply logit soft-cap in the XPU flash-attention kernel

### PR DESCRIPTION
**Pull request by desmond-intel**
## Motivation
The XPU flash-attention `fwd` op accepts a `softcap` argument but never applied it — it was stored in `params.softcap` and the runner structs, yet not forwarded to the softmax collective, and there was no `tanh` in the kernel. So a model's attention logit soft-cap (e.g. Gemma2 `attn_logit_softcapping = 50`) was silently dropped, and Gemma-2 accuracy on `--attention-backend intel_xpu` fell below the triton reference.

## Modifications
- Apply `S = softcap * tanh(S / softcap)` on the natural-scale logit in both the prefill (`FMHAFwdMainloop`) and decode (`DecodeFwdMainloop`) softmax paths — undo the fused `log2(e)` scale before the `tanh`, then re-fold it for the `exp2` softmax. `softcap == 0` keeps the existing fused fast path byte-identical (non-softcap models unaffected).
- Forward `params.softcap` from the prefill and decode runners into the mainloop `Arguments`.

## Accuracy / Tests (Intel Arc Pro B60)
- Softcap probe (identical q/k/v, `softcap` 0 vs 5 vs 50): output now differs (`0↔5` max-diff 3.91) — was byte-identical before.
- `google/gemma-2-27b` hellaswag 5-shot, tp-4, `--attention-backend intel_xpu`: **acc 0.435 / acc_norm 0.554**, matching the triton backend (0.434 / 0.564), up from the broken 0.351 / 0.427.

Pairs with a cicd serve-arg (`--chunked-prefill-size 512`) that addresses gemma-2-27b's separate Level-Zero input-logprob hang; this PR is the accuracy half.

### CODE REVIEW

**Comment by desmond-intel** [comment]
Thanks - all three addressed in the follow-up commit.

**1. Masked-lane leak (masked `-INFINITY` → finite `-softcap`).** Confirmed and fixed. Causal/window/remainder masks set `tSrS = -INFINITY` before `softmax()`, so the cap ran on already-masked scores; `tanh(-inf)` folded them to a finite `-softcap`, letting masked positions (and all-masked rows) leak. The cap lambda now guards `if (softcap > 0 && sycl::isfinite(v))`, leaving non-finite lanes untouched, in both the prefill and decode softmax. Added a regression probe (identical q/k/v, `causal=True`, `softcap=15`) that matches an fp32 softcap-then-mask reference to `max|diff| = 0.009`.

**2. rel_bias + softcap ordering.** Agreed. Since `apply_relative_bias` folds the bias and sets `qk_scale = 1` before softmax, capping there would compute `softcap(QK + bias)` rather than the reference `softcap(QK) + bias`. Both runners now `TORCH_CHECK`-reject the combination (rejected loudly rather than silently wrong; the Gemma-2 target uses no relative bias).

**3. Missing softcap coverage.** Flipped `DISABLE_SOFTCAP = False`, which enables the `softcap = 15` parametrization in `test_flash_attn_varlen_output` across its `causal=True` / `local=True` (masked) shapes — the coverage that exercises the masked interaction from #1.

No accuracy regression: gemma-2-27b hellaswag (5-shot, tp-4, `--attention-backend intel_xpu`) stays acc 0.42 / **acc_norm 0.55**, matching the triton backend.

File: src/sycl/kernels/flash_attention_v2/collective/xe_fmha_fwd_mainloop.hpp [hunk 1]
```diff
@@ -803,22 +805,38 @@ struct FMHAFwdMainloop<
       FragSRow& tS_sum,     // Softmax row-wise sum accumulator
       ElementS qk_scale) {  // Q*K scale (folds in fp8 K per-tensor scale_k)
 
+    // Logit soft-cap (Gemma2 attn_logit_softcapping): cap on the natural-scale logit.
+    // qk_scale folds in log2(e) for the exp2 softmax, so undo it, tanh-cap, re-fold.
+    // softcap == 0 -> identity, so the fused fast path is unchanged for non-softcap models.
+    const ElementS softcap = params.softcap;
+    const auto cap = [softcap](ElementS v) -> ElementS {
```
Comment (Valentine233) [inline src/sycl/kernels/flash_attention_v2/collective/xe_fmha_fwd_mainloop.hpp]: Add softcap as a template argument `Softcap` to use `if constexpr (Softcap)`.

This function is duplicated and we should define a sharing one outside.
```
template <bool HasSoftcap, typename Element>
CUTLASS_DEVICE Element apply_logit_softcap(Element value, Element softcap) {
  if constexpr (HasSoftcap) {
    if (!sycl::isfinite(value)) {
      return value;
    }
    constexpr Element kLog2e = Element(1.4426950408889634074);
    return softcap * sycl::tanh(value / kLog2e / softcap) * kLog2e;
  }
  return value;
}
```

File: tests/test_flash_attention.py [hunk 2]
```diff
@@ -62,7 +62,7 @@ def is_fa3_supported(device=None) -> bool:
 DISABLE_PAGEDKV = False
 DISABLE_APPENDKV = True
 DISABLE_LOCAL = True
-DISABLE_SOFTCAP = True
+DISABLE_SOFTCAP = False  # XPU flash-attn now applies softcap; exercise it (incl. causal/local masked cases)
```
Comment (Valentine233) [inline tests/test_flash_attention.py:65]: This only affects non-paged attn (`test_flash_attn_varlen_output`).
Please also add UTs for paged attn (`test_flash_attn_kvcache`), split-KV (`test_flash_attn_decode_kvcache`) and FP8 path (`test_flash_attn_fp8_kvcache`).

**Review by Valentine233 — CHANGES_REQUESTED** [review]
Please add softcap as a template argument `Softcap`, like Causal and LocalMask, so that we can use `if constexpr (Softcap)` for perf.

**Comment by desmond-intel** [comment]
Thanks @Valentine233, both done.

Made softcap a compile-time template param the same way `Causal`/`LocalMask` work — pulled the cap into a shared `apply_logit_softcap<HasSoftcap>` helper and threaded `Softcap` through the mainloops, the configs and the `.cpp.in` dispatch (`AT_DISPATCH_BOOL_NO_RETURN(params.softcap != 0.f, Softcap, ...)`), so `if constexpr` drops it entirely for the non-softcap kernels and the duplicated lambda is gone. It's behavior-preserving — the softcap and masked-softcap probes are unchanged and gemma-2-27b hellaswag still lands at acc_norm 0.55.

Also parametrized softcap (0 and 15) in the paged, split-KV and fp8 kvcache tests, threading it into both the kernel call and the `attention_ref` reference. Ran a softcap=15 case on each of the three paths — all pass.

File: src/sycl/kernels/flash_attention_v2/collective/xe_fmha_fwd_mainloop.hpp [hunk 3]
```diff
@@ -64,6 +64,22 @@ namespace cutlass::fmha::collective {
...
+// Gemma-style logit soft-cap, applied on the natural-scale logit. qk_scale folds
+// in log2(e) for the exp2 softmax, so undo it, tanh-cap, then re-fold. Non-finite
+// (masked -INFINITY) lanes pass through untouched so masks don't leak. HasSoftcap
+// is compile-time so the non-softcap path is fully elided.
+template <bool HasSoftcap, typename Element>
+CUTLASS_DEVICE Element apply_logit_softcap(Element value, Element softcap) {
+  if constexpr (HasSoftcap) {
+    if (!sycl::isfinite(value)) {
+      return value;
+    }
```
Comment (Valentine233) [inline src/sycl/kernels/flash_attention_v2/collective/xe_fmha_fwd_mainloop.hpp]: Please do `if constexpr (HasSoftcap)` outside the function:
```
if constexpr (HasSoftcap) {
   value = apply_logit_softcap(value, softcap);
}
```

File: src/sycl/kernels/flash_attention_v2/xe_fmha_fwd_decode_fp8_kernel.cpp.in [hunk 4]
```diff
@@ -39,15 +39,25 @@ __attribute__((visibility("default"))) void FmhaDecodeFp8Runner<@QG_SZ@, @HEAD_D
...
+    AT_DISPATCH_BOOL_NO_RETURN(params.softcap != 0.f, Softcap, {
     if (params.is_e4m3) {
       DecodeConfig<
           false, LocalMask, false, /*LSE=*/false, TileShapeQK, TileShapePV, TileShapeOutput, SubgroupLayoutQK, void, 1, false,
-          QElement, cutlass::float_e4m3_t, cutlass::float_e4m3_t, QElement>::run_paged(params);
+          QElement, cutlass::float_e4m3_t, cutlass::float_e4m3_t, QElement,
+          /*PackGQAEnabled=*/true, void,
+          cute::Stride<int, cute::_1, int, int>, cute::Stride<int, cute::_1, int, int>,
+          cute::Stride<cute::_1, int, int, int>, cute::Stride<int, cute::_1, int, int>,
+          void, void, void, void, /*HasRelBias=*/false, Softcap>::run_paged(params);
```
Comment (Valentine233) [inline src/sycl/kernels/flash_attention_v2/xe_fmha_fwd_decode_fp8_kernel.cpp.in]: Please add `Softcap` right before `LSE`, to avoid long param codes. Same for the other places.

**Comment by desmond-intel** [comment]
Done. Moved the `if constexpr (Softcap)` out to the call sites and put `Softcap` before `LSE` in the configs, so the `.cpp.in` just take one extra arg now and I dropped the trailing default lists. Numbers unchanged - probes, the paged/split-KV/fp8 softcap tests, and gemma hellaswag (acc_norm 0.55) all still match.

**Review by Valentine233 — APPROVED** [review]
